### ISSUE TITLE
Extremely slow first epoch.

### ISSUE THREAD

**Issue by w4-kang** [tensorflow/tensorflow#13114]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.2 LTS
- **TensorFlow installed from (source or binary)**: PIP, tensorflow-gpu==1.2.1
- **TensorFlow version (use command below)**: 1.2.1
- **Python version**: Python 3.6.2
- **CUDA/cuDNN version**: CUDA Version 8.0.61 / cuDNN Version 5.1.10
- **GPU model and memory**: GTX 1080ti x 2, 11171MiB each.

### Describe the problem
When training model, "**only First epoch after executing training script is extremely slow** (More specifically, It takes 12 hours while second and third epoch after executed takes 2.5 hours with my environment and dataset)"

I uploaded this issue because similar issue in stackoverflow is not handled long time.
(https://stackoverflow.com/questions/44966831/tensorflow-first-epoch-is-extremely-slow-maybe-related-to-pool-allocator)

In more detail,
my code uses model with  **1d convolution layers** (tf.nn.conv1d - tf.bias_add - tf.nn.relu) as model.
**tf.ctc_loss** as loss function.
Data is served with tf.PaddingFIFOQueue && tf.QueueBase.dequeue_many and "**each data has different size**".

Here is what I tried,
First I also assumed that pool allocator makes this problem like [above stackoverflow link](https://stackoverflow.com/questions/44966831/tensorflow-first-epoch-is-extremely-slow-maybe-related-to-pool-allocator) (**but now I think this issue is not from pool allocator**)
1. Uses tcmalloc
`LD_PRELOAD="/usr/lib/libtcmalloc.so"` which is suggested in [here](https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/ja7FlGrvh-E)
2. Uses BFC allocator type.
- With `TF_CPU_ALLOCATOR_USE_BFC=true`
- And I tried `tf.ConfigProto().gpu_options. allocator_type = 'BFC'` also.
3. Build tensorflow from source with modified initial pool size limit (100 to 10000), (Which is hard coded in [here](https://github.com/tensorflow/tensorflow/blob/927f811b0303e51126531c135f8b093383de2d6d/tensorflow/core/common_runtime/gpu/process_state.cc#L187)) 

And all trying was ineffective.

Except extremely slow first epoch, everything was fine. Weights of model is trained well, save and load checkpoint and continuing training or inferencing is also fine without any warning and error.

Is it one of avoidable characteristic of Tensorflow or bugs?

Thank you for your reading.
If you need any more information, please notify me.

**Comment by ppwwyyxx** [comment]
If each data has different size, TF can spend a large amount of time running cudnn benchmarks for each data and store them in cache. This is probably the reason.
You can `export TF_CUDNN_USE_AUTOTUNE=0` to see if each epoch becomes the same (slower) speed.

**Comment by aselle** [comment]
@w4-kang, does @ppwwyyxx's suggestion improve your performance?

**Comment by aselle** [comment]
Is each data item different size?

**Comment by aselle** [comment]
I would recommend trying to redefine your code in such a way that your data can be same size. That will likely be the best way to avoid performance problems. Make sure you are not modifying the graph continually on every loop of training (you can finalize your graph to get an error if yout ry to do this). Without you sharing your code in the issue (or a simplified version that still has the problem), it's very unlikely we will be able to help you effectively (we're just guessing).

**Comment by w4-kang** [comment]
Even though second epoch is still running, I think elapse time of second epoch will be similar with first epoch(about 45h) with `export TF_CUDNN_USE_AUTOTUNE=0`. As ppwwyyxx 's expectation, two epoch takes same (slower) speed.
If I can't modify data (like zero padding) for like aselle's recomentation, no more options in there?
Rather than make same size data, if I arrange order of data and make each batch group's size similar, does it help this problem?
I haven't tried make all data same size, but when I tried to train with large batch size, there is no significant difference between first & second epoch's elapse time.
And more, When I train with size sorted data, Increment of first epoch's time elapse is significantly smaller than random ordered.

**Comment by aselle** [comment]
Your best bet is probably to sort and to pad within discrete sets of sizes. i.e. one strategy would be to have powers of two sizes allowed with zero padding to the next biggest to bit your irregular sizes. This is edging on the side of a stack overflow question at this point. I think things are working as intended.   We could maybe consider it a feature request to make padding and sorting of input data to be easier (library functions to help). @mrry, do you have any other thoughts?

**Comment by ppwwyyxx** [comment]
Dealing with irregular shapes is quite common. I wish there is a way to use a reasonably good cudnn algorithm without wasting too much time in warm up, but I don't know how this can be done. Maybe allow users to dump/load the autotune cache? Following some heuristics when autotune is disabled?

**Comment by w4-kang** [comment]
Thanks a lot for finding reason of problem. Now, I should find way to dataset reordering method that minimize data modification and  approximatively size sorted. And as person facing with this problem, saving autotune cache persistently might be great like above ppwwyyxx's idea!

**Comment by mrry** [comment]
The NMT tutorial takes a similar approach to batch input data according to bucketed sequence length. You might try taking a look at [how it uses `tf.contrib.data.group_by_window()`](https://github.com/tensorflow/nmt/blob/d43f1a6371e446778d79da991115c68435be9dcd/nmt/utils/iterator_utils.py#L197) to achieve this.

**Comment by w4-kang** [comment]
Thanks for information!
